### PR TITLE
Add a function for the atomic exchange operation

### PR DESCRIPTION
**Pull request by babsingh**
The new `AtomicSupport::lockExchange` function is equivalent to a fetch
and store operation. A new value is stored at a memory location, and the
old value at the memory location is returned.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
**Documentation:**

1. [_InterlockedExchange](https://docs.microsoft.com/en-us/cpp/intrinsics/interlockedexchange-intrinsic-functions?view=vs-2019): Microsoft
2. [__sync_lock_test_and_set](https://gcc.gnu.org/onlinedocs/gcc-4.1.0/gcc/Atomic-Builtins.html): GCC, Clang
3. [__fetch_and_swap](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=2&cad=rja&uact=8&ved=2ahUKEwjA37T2kKHlAhVRxVkKHWfZC6IQFjABegQIAxAE&url=https%3A%2F%2Fwww-01.ibm.com%2Fsupport%2Fdocview.wss%3Fuid%3Dswg27012874%26aid%3D1&usg=AOvVaw3jCY0ofo1k32HDTfUdmKNk): XLC

**Comment by babsingh** [comment]
**Microsoft atomic exchange implementation:**
```
#include <windows.h>
#include <intrin.h>

int main() {
    volatile __int64 * address = 0;
    __int64 newValue = 1;
    _ReadWriteBarrier();
    __faststorefence();
    _ReadWriteBarrier();
    __int64 oldValue = (__int64)_InterlockedExchange64((volatile __int64 *)address, (__int64)newValue);
}
```

**Windows x86 assembly (MSVC v19.22):**
```
newValue$ = 0
address$ = 8
oldValue$ = 16
main    PROC
$LN3:
        sub     rsp, 40                             ; 00000028H
        mov     QWORD PTR address$[rsp], 0
        mov     QWORD PTR newValue$[rsp], 1
        lock or DWORD PTR [rsp], 0
        mov     rax, QWORD PTR newValue$[rsp]
        mov     rcx, QWORD PTR address$[rsp]
        xchg    QWORD PTR [rcx], rax
        mov     QWORD PTR oldValue$[rsp], rax
        xor     eax, eax
        add     rsp, 40                             ; 00000028H
        ret     0
main    ENDP
```

**GNUC atomic exchange implementation:**
```
#include <stdint.h>

int main() {
    volatile uint64_t *address;
    uint64_t newValue = 3;
    __sync_synchronize();
    __sync_lock_test_and_set(address, newValue);
}
```

**x86 gcc7.3 assembly:**
```
        push    rbp
        mov     rbp, rsp
        mov     QWORD PTR [rbp-8], 3
        mfence
        mov     rax, QWORD PTR [rbp-16]
        mov     rdx, QWORD PTR [rbp-8]
        xchg    rdx, QWORD PTR [rax]
        mov     eax, 0
        pop     rbp
        ret
```

 **x86 clang assembly (Apple LLVM version 9.1.0):**
 ```
 main:                                   # @main
		pushq	%rbp
		movq	%rsp, %rbp
		xorl	%eax, %eax
		movq	$3, -16(%rbp)
		mfence
		movq	-8(%rbp), %rcx
		movq	-16(%rbp), %rdx
		xchgq	%rdx, (%rcx)
		movq	%rdx, -24(%rbp)         ## 8-byte Spill
		popq	%rbp
		retq
```

**ARM gcc-7.3.0 assembly:**
```
main:
        push    {r4, fp, lr}
        add     fp, sp, #8
        sub     sp, sp, #20
        mov     r3, #3
        mov     r4, #0
        str     r3, [fp, #-20]
        str     r4, [fp, #-16]
        bl      __sync_synchronize
        ldr     r1, [fp, #-24]
        sub     r3, fp, #20
        ldmia   r3, {r2-r3}
        mov     r0, r1
        bl      __sync_lock_test_and_set_8
        mov     r3, #0
        mov     r0, r3
        sub     sp, fp, #8
        pop     {r4, fp, pc}
```

**ppc64le gcc6.3.0 assembly:**
```
main:
        std 31,-8(1)
        stdu 1,-64(1)
        mr 31,1
        li 9,3
        std 9,32(31)
        sync
        ld 9,40(31)
        ld 10,32(31)
.L2:
        ldarx 8,0,9
        stdcx. 10,0,9
        mcrf 7,0
        bne 7,.L2
        isync
        li 9,0
        mr 3,9
        addi 1,31,64
        ld 31,-8(1)
        blr
        .long 0
        .byte 0,9,0,0,128,1,0,1
```

**s390x gcc-7.4.0 assembly:**
```
        mvghi   160(%r11),3
        bcr     14,0
        lg      %r2,168(%r11)
        lg      %r3,160(%r11)
        lg      %r1,0(%r2)
.L2:
        csg     %r1,%r3,0(%r2)
        jne     .L2
        lhi     %r1,0
        lgfr    %r1,%r1
        lgr     %r2,%r1
        lgdr    %r11,%f2
```

**XLC atomic exchange implementation:**
```
#include <stdlib.h>

int main() {
    volatile uintptr_t *address;
    uintptr_t newValue = 3;
    __sync();
    __fetch_and_swap((volatile int*)address, (int)newValue);
}
```

**ppc64le xlc13.1.3 assembly:**
```
        .csect  H.10.NO_SYMBOL{PR}, 7
.main:                                  # 0x00000000 (H.10.NO_SYMBOL)
        stwu       SP,-80(SP)
        addi       r4,r0,3
        stw        r4,68(SP)
        sync
        lwz        r5,64(SP)
__L14:                                  # 0x00000014 (H.10.NO_SYMBOL+0x14)
        lwarx      r3,r0,r5
        stwcx.     r4,r0,r5
        .machine        "any"
        bc         BO_IF_NOT_3,CR0_EQ,__L14
        addi       r3,r0,0
        addi       SP,SP,80
        bclr       BO_ALWAYS,CR0_LT
        .long   0x00000000
```

**Comment by babsingh** [comment]
@fjeremic I couldn't compile the above XLC implementation (with `__fetch_and_swap`) on zOS. On zLinux (s390x), GCC-7.4 generates a `csg` instruction for the atomic exchange opertion. On zOS, do I need to directly use the `csg` instruction?

File: include_core/AtomicSupport.hpp [hunk 1]
```diff
@@ -438,6 +438,80 @@ class VM_AtomicSupport
...
+	 * Store the unsigned 64-bit value at the memory location as an atomic operation, and
+	 * return the old 64-bit value stored at the memory location.
+	 *
+	 * @param address The memory location to be updated
+	 * @param newValue The new value to be stored at memory address
+	 *
+	 * @return the value at memory location <b>address</b> BEFORE the store was attempted
+	 */
+	VMINLINE static uint64_t
+	lockExchangeU64(volatile uint64_t *address, uint64_t newValue)
```
Comment (fjeremic) [inline include_core/AtomicSupport.hpp]: Looking at these new APIs and the assembly generated, aren't they equivalent to already existing APIs? Isn't this function a copy of `setU64` [1] effectively?

[1] https://github.com/eclipse/omr/blob/3d1baf40fa28352196d77c1a1a07b5c2f7df6145/include_core/AtomicSupport.hpp#L744-L767

**Comment by babsingh** [comment]
@zl-wang fyi https://github.com/eclipse/omr/pull/4469#issuecomment-542880211 contains GCC and XLC assembly for the atomic exchange operation on PPC.

>  I assumed your use case is the latter for MCS. If yes, we have to check EH==0 hint is on the larx (load and reserve) instruction. For exclusive, EH==1 (hardware will hold onto that cacheline longer).

Referring to the above assembly, GCC generates, `ldarx` and XLC generates `lwarx`, for the atomic exchange operation, which are variants of the load and reserve instruction. So, the correct assembly is generated with `__sync_lock_test_and_set` (GCC) and `__fetch_and_swap`(XLC).

> __fence() is ambiguous. ... xlC have __lwsync() --- light-weight sync; and __sync() --- heavy-weight sync for your purpose.

Typo w.r.t. `fence`; actually, meant `__sync()`. Both `__sync_synchronize` (GCC) and `__sync()` (XLC) generate the correct assembly as suggested by you.

**Comment by zl-wang** [comment]
lwarx is for 32bit;  ldarx is for 64bit.   both can have hints (EH=0 or 1).
I need to see the binary  (EH is the LSB --- least significant bit).

__sync() is much more expensive than __lwsync().   the question is which one you need?

File: include_core/AtomicSupport.hpp (hunk 1 above)
Comment (babsingh) [reply include_core/AtomicSupport.hpp]: On x86, `xchg` and `cmpxchg` are different instructions. On PPC, compare-and-swap and exchange yield almost similar assembly. I think the same happens on s390x (CAS == exchange). Should we just call `setU64` from `lockExchange` on PPC and s390x?

**PPC compare-and-swap:**
```
	//__sync_synchronize
	sync
	//__sync_val_compare_and_swap
        ld 9,48(31)
        ld 8,40(31)
        ld 10,32(31)
        sync
        ldarx 7,0,9
        cmpd 7,7,8
        bne 7,.L3
        stdcx. 10,0,9
        mcrf 7,0
        bne 7,.L2
        isync
```

**PPC atomic-exchange:**
```
	//__sync_synchronize
        sync
        //__sync_lock_test_and_set
        ld 9,48(31)
        ld 10,32(31)
        ldarx 8,0,9
        stdcx. 10,0,9
        mcrf 7,0
        bne 7,.L2
        isync
```

Comment (fjeremic) [reply include_core/AtomicSupport.hpp]: For Linux on Z and z/OS yes I think we can just call the `setU32` and `setU64` respectively as they will yield the exact same sequence.

**Comment by babsingh** [comment]
> __sync() is much more expensive than __lwsync(). the question is which one you need?

I think `__sync()` is needed for enforcing a full barrier.

> I need to see the binary (EH is the LSB --- least significant bit).

In the below binary, the LSB (EH) is 1 for `ldarx` (**a8 48 00 7d**), which means exclusive i.e. hardware will hold onto that cache line longer. This looks similar to `setU32` and `setU64`. @zl-wang Would you also suggest using `setU32` and `setU64` in `lockExchange` on PPC?

```
// GNUC implementation of atomic exchange
include <stdint.h>

int main() {
    volatile uint64_t *address;
    uint64_t newValue = 3;
    __sync_synchronize();
    __sync_lock_test_and_set(address, newValue);
}

# gcc -c atomic_exchange_gcc.c -o atomic_exchange_gcc.o
# objdump -d atomic_exchange_gcc.o

atomic_exchange_gcc.o:     file format elf64-powerpcle
Disassembly of section .text:
0000000000000000 <main>:
   0:	f8 ff e1 fb 	std     r31,-8(r1)
   4:	c1 ff 21 f8 	stdu    r1,-64(r1)
   8:	78 0b 3f 7c 	mr      r31,r1
   c:	03 00 20 39 	li      r9,3
  10:	20 00 3f f9 	std     r9,32(r31)
  14:	ac 04 00 7c 	hwsync
  18:	28 00 3f e9 	ld      r9,40(r31)
  1c:	20 00 5f e9 	ld      r10,32(r31)
  20:	a8 48 00 7d 	ldarx   r8,0,r9
  24:	ad 49 40 7d 	stdcx.  r10,0,r9
  28:	00 00 80 4f 	mcrf    cr7,cr0
  2c:	f4 ff 9e 40 	bne     cr7,20 <main+0x20>
  30:	2c 01 00 4c 	isync
  34:	78 4b 23 7d 	mr      r3,r9
  38:	40 00 3f 38 	addi    r1,r31,64
  3c:	f8 ff e1 eb 	ld      r31,-8(r1)
  40:	20 00 80 4e 	blr
	...
  4c:	80 01 00 01 	.long 0x1000180
```

**Comment by zl-wang** [comment]
@babsingh  as your use-case is atomic update to build/dismantle the MCS chain (not exclusive), the EH hint used should be 0. so, __sync_lock_test_and_set() looks like not the right primitive for your purpose. 

I supposed xlC built-in __compare_and_swaplp and GCC built-in __sync_val_compare_and_swap are the right thing to use (so that setU64).  Check out the binary to be sure ...

**Comment by babsingh** [comment]
> __compare_and_swaplp and GCC built-in __sync_val_compare_and_swap are the right thing to use

Correct. They have EH set to 0 (`a8 48 e0 7c 	ldarx   r7,0,r9`). I will use setU32/setU64 to build/dismantle the MCS queue.

Found this Intel Article [Implementing Scalable Atomic Locks ...](https://software.intel.com/en-us/articles/implementing-scalable-atomic-locks-for-multi-core-intel-em64t-and-ia32-architectures/), which states that
> One common myth is that the lock utilizing a cmpxchg instruction is cheaper than a lock utilizing an xchg instruction. This is used because cmpxchg will not attempt to get the lock in exclusive mode since the cmp will go through first. Figure 9 shows that the cmpxchg is just as expensive as the xchg instruction.

Assuming that the performance of fetch-and-store and compare-and-swap varies between architectures (x86, PPC and S390x). I feel that OMR should support both fetch-and-store and compare-and-swap as a feature.

**Comment by fjeremic** [comment]
Waiting for an approval record from @zl-wang before merging.

**Pull request merged**